lib/Wasmtime.hs

```diff
@@ -952,43 +967,36 @@ newExternTypeFromPtr externtype_ptr = do
 -- number of instances in it because 'Store' will never release this
 -- memory. It’s recommended to have a 'Store' correspond roughly to the lifetime
 -- of a “main instance” that an embedding is interested in executing.
-newtype Store s = Store {unStore :: ForeignPtr C'wasmtime_store_t}
+data Store s = Store
+  { -- | A mutable finalizer which is used to finalize FunPtrs of
+    -- FuncCallbacks. See 'newFunc' where this finalizer is extended. This
+    -- finalizer is run in the finalizer of the 'storeForeignPtr' in 'newStore'
+    -- below.
+    storeFinalizeRef :: !(IORef (IO ())),
+    storeForeignPtr :: !(ForeignPtr C'wasmtime_store_t),
+    storeCtxPtr :: !(Ptr C'wasmtime_context_t)
+  }
 
 -- | Creates a new store within the specified engine.
 newStore :: MonadPrim s m => Engine -> m (Store s)
 newStore engine = unsafeIOToPrim $ withEngine engine $ \engine_ptr -> mask_ $ do
   wasmtime_store_ptr <- c'wasmtime_store_new engine_ptr nullPtr nullFunPtr
   checkAllocation wasmtime_store_ptr
-  Store <$> newForeignPtr p'wasmtime_store_delete wasmtime_store_ptr
-
-withStore :: Store s -> (Ptr C'wasmtime_store_t -> IO a) -> IO a
-withStore = withForeignPtr . unStore
-
--- | A context is basically the same as a 'Store'. All functions in @wasmtime@
--- that require a 'Store' will actually require a 'Context' instead.
---
--- We might fold 'Store' and 'Context' into a single type in the future.
-data Context s = Context
-  { -- | Usage of a @wasmtime_context_t@ must not outlive the original @wasmtime_store_t@
-    -- so we keep a reference to a 'Store' to ensure it's not garbage collected.
-    storeContextStore :: !(Store s),
-    storeContextPtr :: !(Ptr C'wasmtime_context_t)
-  }
-
--- | Get a 'Context' from a 'Store'.
-storeContext :: MonadPrim s m => Store s -> m (Context s)
-storeContext store =
-  unsafeIOToPrim $ withStore store $ \wasmtime_store_ptr -> do
-    wasmtime_ctx_ptr <- c'wasmtime_store_context wasmtime_store_ptr
-    pure
-      Context
-        { storeContextStore = store,
-          storeContextPtr = wasmtime_ctx_ptr
-        }
-
-withContext :: Context s -> (Ptr C'wasmtime_context_t -> IO a) -> IO a
-withContext ctx f = withStore (storeContextStore ctx) $ \_store_ptr ->
-  f $ storeContextPtr ctx
+  wasmtime_ctx_ptr <- c'wasmtime_store_context wasmtime_store_ptr
+  finalizeRef <- newIORef $ pure ()
+  storeFP <- Foreign.Concurrent.newForeignPtr wasmtime_store_ptr $ do
+    c'wasmtime_store_delete wasmtime_store_ptr
+    join $ readIORef finalizeRef
+  pure
+    Store
+      { storeFinalizeRef = finalizeRef,
+        storeCtxPtr = wasmtime_ctx_ptr,
+        storeForeignPtr = storeFP
+      }
+
+withStore :: Store s -> (Ptr C'wasmtime_context_t -> IO a) -> IO a
+withStore store f = withForeignPtr (storeForeignPtr store) $ \_store_ptr ->
```

Comment:
why the withForeignPtr here, if we don't use it? does this 'keepalive' the store?

Reply:
Yes it keeps the store alive.